File: crates/chezmoi-support/src/vscode.rs
use crate::command::{command_text, run_command};
use crate::context::{Options, context_with_options};
use crate::error::Result;
use serde::Deserialize;
use std::path::Path;

#[derive(Debug, Deserialize)]
struct ExtensionManifest {
    extensions: Vec<ExtensionSpec>,
}

#[derive(Debug, Deserialize)]
struct ExtensionSpec {
    id: String,
}

pub fn install_vs_extensions(options: &Options) -> Result<()> {
    if which::which("code").is_err() {
        return Ok(());
    }
    let ctx = context_with_options(options)?;
    let extensions_file = ctx
        .source_dir
        .join("dot_config/Code/User/vscode-extensions.toml");
    if !extensions_file.exists() {
        return Ok(());
    }
    let installed = command_text(&duct::cmd("code", ["--list-extensions"]))?;
    for extension in extension_ids(&extensions_file)? {
        if installed
            .lines()
            .any(|line| line.trim().eq_ignore_ascii_case(&extension))
        {
            continue;
        }
        run_command(&duct::cmd(
            "code",
            ["--install-extension", &extension, "--force"],
        ))?;
    }
    Ok(())
}

fn extension_ids(path: &Path) -> Result<Vec<String>> {
    let manifest: ExtensionManifest = toml::from_str(&fs_err::read_to_string(path)?)?;
    Ok(manifest
        .extensions
        .into_iter()
        .map(|extension| extension.id.trim().to_owned())
        .filter(|extension| !extension.is_empty())
        .collect())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_extensions_from_toml() -> Result<()> {
        let temp = tempfile::NamedTempFile::new()?;
        fs_err::write(
            temp.path(),
            "[[extensions]]\nid = \"one.alpha\"\n\n[[extensions]]\nid = \"\"\n\n[[extensions]]\nid = \" Two.Beta \"\n",
        )?;

        assert_eq!(
            extension_ids(temp.path())?,
            vec!["one.alpha".to_owned(), "Two.Beta".to_owned()]
        );
        Ok(())
    }
}

File: internal/chezmoi/vscode.go
package chezmoi

import (
	"errors"
	"os"
	"path/filepath"
	"strings"

	"github.com/FlameFlag/nix-dotfiles/internal/common/process"
	"github.com/pelletier/go-toml/v2"
)

type extensionManifest struct {
	Extensions []extensionSpec `toml:"extensions"`
}

type extensionIDManifest struct {
	Extensions []string `toml:"extensions"`
}

type extensionSpec struct {
	ID string `toml:"id"`
}

func InstallVSExtensions(options Options) error {
	if _, ok := process.PathOf("code"); !ok {
		return nil
	}
	ctx, err := ContextWithOptions(options)
	if err != nil {
		return err
	}
	extensionsFile := filepath.Join(ctx.SourceDir, "dot_config/Code/User/vscode-extensions.toml")
	if _, err := os.Stat(extensionsFile); errors.Is(err, os.ErrNotExist) {
		return nil
	}
	installed, err := commandText([]string{"code", "--list-extensions"})
	if err != nil {
		return err
	}
	for _, extension := range extensionIDsFromFile(extensionsFile) {
		alreadyInstalled := false
		for line := range strings.SplitSeq(installed, "\n") {
			if strings.EqualFold(strings.TrimSpace(line), extension) {
				alreadyInstalled = true
				break
			}
		}
		if alreadyInstalled {
			continue
		}
		if err := process.RunInWithEnvAndStdin(
			"",
			[]string{"code", "--install-extension", extension, "--force"},
			nil,
			os.Stdin,
		); err != nil {
			return err
		}
	}
	return nil
}

func extensionIDsFromFile(path string) []string {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}

	var idManifest extensionIDManifest
	if err := toml.Unmarshal(data, &idManifest); err == nil {
		return cleanExtensionIDs(idManifest.Extensions)
	}

	var manifest extensionManifest
	if err := toml.Unmarshal(data, &manifest); err != nil {
		return nil
	}
	var ids []string
	for _, extension := range manifest.Extensions {
		ids = append(ids, extension.ID)
	}
	return cleanExtensionIDs(ids)
}

func cleanExtensionIDs(ids []string) []string {
	var out []string
	for _, id := range ids {
		id = strings.TrimSpace(id)
		if id != "" {
			out = append(out, id)
		}
	}
	return out
}
